applications/nrf_desktop/src/util/dfu_lock.c

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/atomic.h>
+
+#include <zephyr/logging/log.h>
+#include <zephyr/logging/log_ctrl.h>
+LOG_MODULE_REGISTER(dfu_lock, CONFIG_DESKTOP_DFU_LOCK_LOG_LEVEL);
+
+#include "dfu_lock.h"
+
+static const struct dfu_lock_owner *previous_owner;
+static const struct dfu_lock_owner *current_owner;
+
+static atomic_t dfu_lock = ATOMIC_INIT(false);
+
+bool dfu_lock_try(const struct dfu_lock_owner *new_owner)
```

Comment:
I wonder if flash accesses shouldn't actually go through the locking utility (so that we would actually enforce proper API usage)? That would make solution far less error-prone.
We could either gate read/write/erase or maybe at least opening and closing the flash area.